dra scheduler: ensure that we never have nil claim/class parameters

klueska opened the pull request:
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
This PR adds logic to define **default** claim/class parameters that will allow
claim allocation to proceed even if an end user doesn't provide claim
parameters themselves.

Without this change, the scheduler was crashing in newClaimController() in `pkg/scheduler/framework/plugins/dynamicresources/structuredparameters.go`

The code in newClaimController() assumes that claim parameters are not nil.
Furthermore it assumes that there is at least one DriverRequest populated in
order to allocate any resources to a claim.

However, we shouldn't *force* a user to supply a ResourceClaimParameters
object if they don't want to. We also shouldn't force a driver implementor to
write a controller just to associate a static ResourceClaimParameters object
with every claim.

In this PR, the "default" claim parameters objects are hard-coded in the
scheduler code. This won't work going forward as we add new structured
models though. What we probably want is a field in the ResourceClass which
points to a default ResourceClaimPrameters object that should be used if one
is not supplied as part of the claim.

```release-note
None
```

klueska committed:
dra scheduler: create default claim/class parameters instead of nil

Without this, the scheduler was crashing in newClaimController() in
pkg/scheduler/framework/plugins/dynamicresources/structuredparameters.go

The code in newClaimController() assumes that the parameters are not nil.
Furthermore it assumes that there is at least one DriverRequest populated in
order to allocate any resources to a claim.

This PR adds logic to define default claim/class parameters that will allow
allocation to proceed even if an end user doesn't provide any class or claim
parameters themselves.

Signed-off-by: Kevin Klues <kklues@nvidia.com>

klueska commented:
/assign @pohly

klueska commented:
I'm realizing this is not sufficient. I think we need a fully populated "default" claim parameters object that requests a single resource with no configuration parameters.

klueska commented:
I have currently hard-coded a "default" claim parameters object in the scheduler code. This won't work going forward as we add new structured models though. What we probably want is a field in the `ResourceClass` which points to a *default* `ResourceClaimPrameters` object that should be used if one is not supplied as part of the claim.

pohly commented:
> This PR adds logic to define default claim/class parameters that will allow
claim allocation to proceed even if an end user doesn't provide claim
parameters themselves.

I wasn't sure whether we want that. The E2E tests use `selector: "true"`, just as in your default. But you are right, this is unnecessarily complex for users. And of course it shouldn't crash... :sweat_smile: 

> What we probably want is a field in the ResourceClass which points to a default ResourceClaimPrameters object that should be used if one is not supplied as part of the claim.

... because without claim parameters, the scheduler doesn't know the claim is for a resource using model "named resources" or some future "some-other-resources". Makes sense.

As a stop-gap solution this is okay.

/lgtm
/assign @alculquicondor 

For approval.

ravisantoshgudimetla commented:
I'm looking into implementing a custom driver and noticed this issue. I like the defaulting idea, if you guys are fine with it, I can take care of it.

pohly commented:
@klueska: we now have the second commit from https://github.com/kubernetes/kubernetes/pull/123831 also in this PR. Can you take it out again?

klueska commented:
I reversed the order so that the other one can merge without further approval. It just needs a new LGTM

pohly commented:
@ravisantoshgudimetla: let's discuss in https://github.com/kubernetes/kubernetes/issues/123858.

klueska committed:
dra kubelet: fix error log

Previously we were returning the error string from 'err' (which is nil), when
we should have been returning it from result.Error. Without this it is hard to
debug issues with NodeUnprepareResources.

Signed-off-by: Kevin Klues <kklues@nvidia.com>

alculquicondor commented on pkg/scheduler/framework/plugins/dynamicresources/structuredparameters.go:
```diff
@@ -128,12 +128,10 @@ func newClaimController(logger klog.Logger, class *resourcev1alpha2.ResourceClas
+				break
 			}
 		}
 		controller, err := namedresourcesmodel.NewClaimController(filter, perDriver.requests)
```
Was the existing behavior equivalent in this case?

pohly replied:
It looks the same to me. `classParameters` is now guaranteed to not be nil (a case that I had considered, in contrast to nil claim parameters), so the if check is not needed anymore. If it had been nil before, it's now empty, so the for loop doesn't do anything, just as before.

alculquicondor commented on pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:
```diff
@@ -1004,9 +1006,31 @@ func (pl *dynamicResources) lookupClassParameters(logger klog.Logger, class *res
+							// match the default to how the resources of this
+							// class are being advertized in a ResourceSlice.
+							NamedResources: &resourcev1alpha2.NamedResourcesRequest{
+								Selector: "true",
```
I don't like that we would have to parse this selector. I would imagine that a case of "no parameters" is fairly common, so we should optimize for by understanding `nil`.

klueska replied:
The problem is that the scheduler doesn't inherently know which resource request model a driver is using to advertise its resources. We only know this by parsing this object and seeing which field is set (in this case `NamedResources`).

As such, I think a better approach would be to allow a cluster admin to define a default ResourceClaimParameters object and link it to a resource class. However, this is a larger chunk of work (and we are past the code freeze), so I proposed adding this PR as a stop-gap for 1.30.

alculquicondor replied:
> The problem is that the scheduler doesn't inherently know which resource request model a driver is using.

Does it matter? The point is that it passes the "filter", regardless of the model.

klueska replied:
I'll let @pohly comment on that as I'm not intimately familiar with all of the code in this plugin. All I know is *some* default should be set and the default should allow sharing. Whether that's through an explicit object or a branch off of detecting `nil` I'll leave up to Patrick.

pohly replied:
I like the approach in this PR better because the rest of the code currently expects to determine the model based and thus the resources based on the parameters. Making it handle nil adds more special cases.

There's also a clear path towards what we really want (configurable default parameters). We need those because detecting the model may not be enough, we also need to figure out what the filters for a claim with no parameters should be. This depends on the driver and/or cluster, not the model.

https://github.com/kubernetes/kubernetes/issues/123858 is tracking this next step.

alculquicondor commented:
/lgtm
/approve
Please have a test case for this before the freeze

pohly commented:
I am adding an E2E test for this. Let's prioritize adding the fix first.

pohly commented:
/cc @kubernetes/release-team 

Another bug fix for a new feature in 1.30, please add the milestone.

katcosgrove commented:
/milestone v1.30
